### PR TITLE
[FLINK-16977][docs] Add docs for components and distributed architecture

### PR DESCRIPTION
**Pull request by StephanEwen**
This adds documentation for the components of a distributed deployment.

image

image

### CODE REVIEW

File: docs/concepts/distributed_architecture.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+A Stateful Functions deployment consists of a few components interacting together. Here we describe these pieces and their
```
Comment (sjwiesman) [inline docs/concepts/distributed_architecture.md:27]: You're missing the second half of this sentence.

File: docs/concepts/distributed_architecture.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+An alternative way of deploying functions is *co-location* with the Flink JVM processes. In such a setup, each Flink TaskManager would talk to one Function process sitting *"next to it"*. A common way to do this is to use a system like Kubernetes and deploy pods consisting of a Flink container and the function side-car container; the two communicate via the pod-local network.
+
+This mode supports different languages while avoiding to route invocations through a Service/LoadBalancer, but it cannot scale the state and compute parts independently.
+
+<p class="text-center">
+	<img width="80%" src="{{ site.baseurl }}/fig/concepts/arch_funs_colocated.svg"/>
+</p>
+
+This style of deployment is similar to how Flink's Table API and API Beam's portability layer and deploy execute non-JVM functions.
```
Comment (sjwiesman) [inline docs/concepts/distributed_architecture.md:89]:
```suggestion
This style of deployment is similar to how Flink's Table API and API Beam's portability layer deploy and execute non-JVM functions.
```

File: docs/concepts/distributed_architecture.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+## High-level View
+
+A *Stateful Functions* deployment consists of a set of **Apache Flink Stateful Functions** processes and, optionally, various deployments that execute remote functions.
+
+<p class="text-center">
+    <img width="80%" src="{{ site.baseurl }}/fig/concepts/arch_overview.svg"/>
+</p>
+
+The Flink worker processes (TaskManagers) receive the events from the ingress systems (Kafka, Kinesis, etc.) and route them to the target functions. They invoke the functions, and route the resulting mesages to the next respective target functions. Messages designated for egress are written to an egress system (again, Kafka, Kinesis, ...).
```
Comment (sjwiesman) [inline docs/concepts/distributed_architecture.md:40]:
```suggestion
The Flink worker processes (TaskManagers) receive the events from the ingress systems (Kafka, Kinesis, etc.) and route them to the target functions. They invoke the functions and route the resulting messages to the next respective target functions. Messages designated for egress are written to an egress system (again, Kafka, Kinesis, ...).
```

**Comment by tzulitai** [comment]
Addressing comments and merging this!

File: docs/concepts/distributed_architecture.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
+---
+title: Distributed Architecture 
+nav-id: dist-arch
+nav-pos: 2
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply docs/concepts/distributed_architecture.md:4]: I agree here, for me personally the flow is better that way.

File: docs/concepts/distributed_architecture.md (hunk 1 above)
Comment (sjwiesman) [reply docs/concepts/distributed_architecture.md:27]: Here we describe these pieces and their relationship to each other and the Apache Flink runtime.

**Comment by tzulitai** [comment]
LGTM, will address the comments while merging!

**Pull request closed without merging**